lib/controllers/login.js

```diff
@@ -4,152 +4,168 @@ var _ = require('lodash');
-  var formActionUri = (config.web.login.uri + (nextUri ? ('?next=' + nextUri) : ''));
-  var view = req.app.get('stormpathConfig').web.login.view;
-  var oauthStateToken = null;
-  var providerData = req.body && req.body.providerData;
-
-  function renderForm(form, options) {
-    if (options === undefined) {
-      options = {};
-    }
-
-    var hasSocialProviders = _.some(config.socialProviders, function (socialProvider) {
-      return socialProvider.enabled;
-    });
 
-    extend(options, {
-      form: form,
-      formActionUri: formActionUri,
-      oauthStateToken: oauthStateToken,
-      hasSocialProviders: hasSocialProviders
-    });
+  credentials = credentials || {};
+  credentials.username = credentials.login || credentials.username || '';
+  credentials.password = credentials.password || '';
 
-    helpers.render(req, res, view, options);
+  if (!credentials.username || !credentials.password) {
+    return callback(new Error('Invalid username or password.'));
   }
 
-  // If a GET request is made with the Accept header as json,
-  // respond with the view model for this form.
-  // TODO: When `stormpath.web.produces` is implemented,
-  // this `if` must check that it is set to `application/json`.
-  if (req.method === 'GET' && accepts === 'json') {
-    return helpers.getLoginViewModel(config, function (err, viewModel) {
+  authenticator.authenticate(credentials, function (err, authenticationResult) {
+    if (err) {
+      return callback(err);
+    }
+
+    authenticationResult.getAccount(function (err, account) {
       if (err) {
-        return res.status(400).json({ error: err.userMessage || err.message });
+        return callback(err);
       }
 
-      res.status(200).json(viewModel);
+      callback(null, {
+        account: account,
+        authenticationResult: authenticationResult
+      });
     });
-  }
+  });
+}
 
-  if (req.user && config.web.login.enabled) {
-    var nextUrl = nextUri || config.web.login.nextUri;
-    return res.redirect(302, nextUrl);
-  }
+/**
+ * This controller logs in an existing user.  If there are any errors, an
+ * error page is rendered.  If the process succeeds, the user will be logged in
+ * and redirected.
+ *
+ * @method
+ *
+ * @param {Object} req - The HTTP request.
+ * @param {Object} res - The HTTP response.
+ * @param {function} next - The next function.
+ */
+module.exports = function (req, res, next) {
+  var config = req.app.get('stormpathConfig');
+  var application = req.app.get('stormpathApplication');
 
   res.locals.status = req.query.status;
 
-  if (req.method === 'POST' && accepts === 'json') {
-    if (providerData) {
-      // Social Login
-      var options = {
-        providerData: {
-          providerId: providerData.providerId,
-          accessToken: providerData.accessToken,
-          code: providerData.code
-        }
-      };
+  helpers.handleContentTypeRequest(req, {
+    'application/json': function () {
```

Comment:
Hey guys, I appreciate the proactive refactoring.  However, these types of changes should happen outside of a feature sprint (the 2.5 milestone we are working on).  If we are going to introduce HTTP abstractions we need to design those as a team, so that everyone is on board.  Unfortunately there just isn’t time for this right now, as we need to get this 2.5 release out ASAP.  As such, please do not introduce any more HTTP abstractions as this time.  In this particular case, our previous approach of using the `req.accepts()` API, with simple if statements, is working perfectly fine.